platform/security/idp/security-idp-plugin/security-idp-plugin-nameidpolicy/src/main/java/org/codice/ddf/security/idp/plugin/nameidpolicy/NameIdPolicyPresignPlugin.java

```diff
@@ -0,0 +1,136 @@
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.security.idp.plugin.nameidpolicy;
+
+import ddf.security.samlp.SamlProtocol.Binding;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.commons.lang.StringUtils;
+import org.codice.ddf.security.idp.plugin.SamlPresignPlugin;
+import org.opensaml.core.xml.schema.XSString;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.NameID;
+import org.opensaml.saml.saml2.core.NameIDPolicy;
+import org.opensaml.saml.saml2.core.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This plugin is responsible for setting the NameID Format/SPNameQualifier on the outgoing Response
+ * according to the values found in the NameIDPolicy of the passed in AuthnRequest.
+ */
+public class NameIdPolicyPresignPlugin implements SamlPresignPlugin {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(NameIdPolicyPresignPlugin.class);
+
+  private static final String EMAIL_ATTRIBUTE_NAME =
+      "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress";
+
+  @Override
+  public void processPresign(
+      Response response,
+      AuthnRequest authnRequest,
+      List<String> spMetadata,
+      Set<Binding> supportedBindings) {
+    NameIDPolicy nameIdPolicy = authnRequest.getNameIDPolicy();
+
+    if (nameIdPolicy == null
+        || (StringUtils.isEmpty(nameIdPolicy.getFormat())
+        && StringUtils.isEmpty(nameIdPolicy.getSPNameQualifier()))) {
+      return;
+    }
+
+    response
+        .getAssertions()
+        .stream()
+        .map(assertion -> assertion.getSubject().getNameID())
+        .forEach(
+            assertionNameId ->
+                setAssertionNameIdQualifierAndFormat(
+                    assertionNameId,
+                    nameIdPolicy.getFormat(),
+                    nameIdPolicy.getSPNameQualifier(),
+                    response));
+  }
+
+  private void setAssertionNameIdQualifierAndFormat(
+      NameID assertionNameId,
+      @Nullable String nameIdFormatPolicy,
+      @Nullable String spNameQualifierPolicy,
+      Response response) {
+
+    if (StringUtils.isNotEmpty(spNameQualifierPolicy)) {
+      assertionNameId.setSPNameQualifier(spNameQualifierPolicy);
+    }
+
+    if (StringUtils.isNotEmpty(nameIdFormatPolicy)) {
+      switch (nameIdFormatPolicy) {
+        // supported NameIDFormats
+        case NameID.UNSPECIFIED:
+          break;
+        case NameID.PERSISTENT:
+          // TODO DDF-3965
+          break;
+        case NameID.TRANSIENT:
```

Comment:
❗️ We technically don't support this in our metadata, and I don't think STS will ever give us a NameID in this format.

Reply:
Should I take out this option, and just return an error if we get the `transient` NameIDPolicy Format?

If we keep it in, it shouldn't cause any problems. It would just be a little more permissive.

Reply:
I think it should be removed. I don't think this would impact the DDF SP as it is now, but I think it would be surprising and cause odd behavior for any other SP that is getting a "transient" identifier that isn't actually transient.